polkadot/xcm/procedural/src/builder_pattern.rs

```diff
@@ -0,0 +1,93 @@
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use syn::{parse_macro_input, DeriveInput, Data, Error, Fields};
+use quote::{quote, format_ident};
+
+pub fn derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+	let builder_impl = match &input.data {
+		Data::Enum(data_enum) => generate_methods_for_enum(input.ident, data_enum),
+		_ => return Error::new_spanned(&input, "Expected the `Instruction` enum")
+			.to_compile_error().into(),
+	};
+	let output = quote! {
+		pub struct XcmBuilder<Call>(Vec<Instruction<Call>>);
+		impl<Call> Xcm<Call> {
+			pub fn builder() -> XcmBuilder<Call> {
+				XcmBuilder::<Call>(Vec::new())
+			}
+		}
+		#builder_impl
+	};
+	output.into()
+}
+
+fn generate_methods_for_enum(name: syn::Ident, data_enum: &syn::DataEnum) -> TokenStream2 {
+	let methods = data_enum.variants.iter().map(|variant| {
+		let variant_name = &variant.ident;
+		let method_name_string = to_snake_case(&variant_name.to_string());
+		let method_name = syn::Ident::new(&method_name_string, variant_name.span());
+		match &variant.fields {
+			Fields::Unit => {
+				quote! {
+					pub fn #method_name(mut self) -> Self {
```

Comment:
What would be really amazing is if you can pull the docs for the individual XCM instructions to be above this fn item. Should not be hard to do.